docs(docs-infra): drop tabs layout from the API reference details page

hawkgs opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

The contents of the API reference details page are separated by tabs (Material).

## What is the new behavior?

The tabs layout is removed in favor of a single page separated by sections. This resulted in the following major changes:

- Removal of Material tabs
- Refactoring of the API gen pipeline rendering templates to match the new tabless layout, i.e. the `tab-*.tsx` templates had been updated to `section-*.tsx`. There are some other minor changes like removal of redundant wrapping elements, etc.
- Reducing the template of the API and CLI reference details page to a single `docs-viewer` call
- Clean up of the scroll services (no longer needed code)
- Refactoring of the reference details styles – now they reside in a `shared-docs/styles` since they are applied to `docs-viewer`
- Added handler for legacy `?tab=<TAB>` URL-s that scroll you to the section that represents the tab

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

hawkgs committed:
docs(docs-infra): drop tabs layout from the API reference details page

Drop the tabs in favor of a single page separated by sections.

kirjs commented:
It seems like there are some infra issues, can you please rebase to see if it helps?

hawkgs commented:
Done.

thePunderWoman commented on adev/src/app/features/references/cli-reference-details-page/cli-reference-details-page.component.ts:
```diff
@@ -61,26 +44,7 @@ export default class CliReferenceDetailsPage implements OnInit {
       )
       .subscribe((doc: DocContent | undefined) => {
-        this.setContentForPageSections(doc);
+        this.docContent.set(doc);
```
Rather than using a setter, you could use the `toSignal` helper, since that's essentially what you're doing here.

hawkgs replied:
Yes💡☝️. Totally missed that while updating the line.

thePunderWoman commented on adev/src/app/features/references/api-reference-details-page/api-reference-details-page.component.ts:
```diff
@@ -6,113 +6,50 @@
+        // Hence, the usage of `setTimeout`.
+        setTimeout(() => {
+          section.scrollIntoView({behavior: 'smooth'});
+        }, 100);
```
You can probably shorten this timeout. We often will do a timeout of 0 for triggering change detection. You could try that.

hawkgs replied:
This is what I did initially but it didn't work (`setTimeout` without timeout to be precise, so it should've been a few ms). That's why I gave it a bit of leeway.

thePunderWoman commented on adev/src/app/features/references/api-reference-details-page/api-reference-details-page.component.spec.ts:
```diff
@@ -6,37 +6,29 @@
-  <div data-tab="Usage Notes" data-tab-url="usage-notes" class="adev-reference-tab"></div>
-  <div class="docs-reference-members-container"></div>
-</div>`;
+  const SAMPLE_CONTENT_WITH_TABS = `<div class="docs-api">
```
Nit: Given this is no longer a set of tabs, consider renaming the variable to reflect that.

hawkgs commented:
It's possible that the change resulted in a flaky unit test. Though, not sure if it's valid for `main` as well, TBH. I'll look into it tomorrow.

thePunderWoman commented:
> It's possible that the change resulted in a flaky unit test. Though, not sure if it's valid for `main` as well, TBH. I'll look into it tomorrow.

It looked like a flake to me. I re-ran the test.

kirjs commented on adev/shared-docs/pipeline/api-gen/rendering/templates/section-heading.tsx:
```diff
@@ -0,0 +1,25 @@
+
+  return (
+    <h2 id={id} class={SECTION_HEADING}>
+      <a href={'#' + id} aria-label={label} tabIndex={-1}>
```
tabIndex should now be sectionIndex. 
Haha j/k please ignore

hawkgs commented on adev/shared-docs/components/viewers/docs-viewer/docs-viewer.component.ts:
```diff
@@ -341,7 +341,13 @@ export class DocViewer implements OnChanges {
+          // Unless this is a link to an element within the same page,
+          // use the Angular router.
+          // https://github.com/angular/angular/issues/30139
+          const scrollToElement = relativeUrl.replace(this.location.path(), '').startsWith('#');
```
Yes, it can. :+1: I don't really use the element ID apart from the check, it seems.

hawkgs commented on adev/shared-docs/pipeline/api-gen/rendering/templates/class-member.tsx:
```diff
@@ -51,20 +51,16 @@ export function ClassMember(props: {member: MemberEntryRenderable}) {
+        <h3>{memberName}</h3>
+        {isClassMethodEntry(props.member) && props.member.signatures.length > 1 ? (
+          <span>{props.member.signatures.length} overloads</span>
+        ) : returnType ? (
```
Sure, it's not a problem.

hawkgs commented on adev/shared-docs/pipeline/api-gen/rendering/transforms/reference-section-id.ts:
```diff
@@ -0,0 +1,14 @@
+  return sectionName
+    .toLowerCase()
+    .replace(/\s|\//g, '-') // remove spaces and slashes
+    .replace(/[^0-9a-zA-Z\-]/g, ''); // only keep letters, digits & dashes
```
Yep. It's another old code snippet that I haven't paid attention to.

hawkgs commented on adev/src/app/features/references/services/reference-scroll-handler.service.ts:
```diff
@@ -33,28 +26,10 @@ export class ReferenceScrollHandler {
 
   private setupCodeToCListeners(tocSelector: string): void {
-    const tocContainer = this.document.querySelector<HTMLDivElement>(tocSelector);
+    const tocContainer = this.document.querySelector<HTMLDivElement>(`.${tocSelector}`);
```
You are right :+1:

hawkgs committed:
fixup! docs(docs-infra): drop tabs layout from the API reference details page

- Address reviewers' comments.
- Use router input binding for the doc content instead of `toSignal` + route data observable.
- Use self-closing tags for docs-viewer

pkozlowski-opensource commented on adev/src/app/features/references/cli-reference-details-page/cli-reference-details-page.component.ts:
```diff
@@ -6,81 +6,23 @@
 
-    element.remove();
-  }
+  docContent = toSignal<DocContent>(
```
can we us input binding instead `toSignal`?

thesmiler commented:
looks like the check didn't like the double fixup! - can you remove one from the last fixup commit?

hawkgs commented:
@thePunderWoman, @kirjs, @pkozlowski-opensource, is there anything else in the diff that you want me to change or improve in some way?

pkozlowski-opensource commented on adev/src/app/features/references/api-reference-details-page/api-reference-details-page.component.html:
```diff
@@ -1,30 +1,9 @@
+    [docContent]="docContent.contents"
+    [hasToc]="true"
+    (contentLoaded)="onContentLoaded()"
+  ></docs-viewer>
```
nit: self-closing tag?

thePunderWoman commented:
This PR was merged into the repository by commit b6733eeea48a86c928ed2f838eddedbfb92e1e12.

The changes were merged into the following branches: main, 19.1.x
